### ISSUE TITLE
uhppoted-mqtt doesn't restart after power outage

### ISSUE THREAD

**Issue by YBonline** [uhppoted/uhppoted#15]
There seems to be a lockfile issue on Ubuntu 22.04, uhppoted-mqtt won't startup from the startup script and fails somewhat silently. I was able to resolve the issue by manually deleting /var/uhppoted/uhppoted.lock

It also seems redundant to the uhppoted-mqtt.pid file, which the start-stop script seems to handle without issue to prevent you from starting it multiple times?

**Comment by uhppoted** [comment]
I _think_ the lockfile you're referring to is the MQTT client lockfile - is the `mqtt.client.id` set to `uhppoted` in your _uhppoted.conf_?

It serves a different purpose to the _uhppoted-mqtt.pid_ - the latter is the lockfile for the _uhppoted-mqtt_ application/service whereas the _uhppoted.lock_ file is to avoid some bizarre edge cases with multiple MQTT client connections using the same client ID which leads to clients being continually disconnected and reconnected.

In theory it gets cleaned up when _uhppoted-mqtt_ shuts down but on a power outage, all bets are off - let me see if I can find a more robust alternative. 

In the interim, you could try adding an `ExecStartPre` hook to the service file (normally _/etc/systemd/system/uhppoted-mqtt.service_)- something like:
```
...
...
[Service]
ExecStartPre=rm -f /var/uhppoted/uhppoted.lock
ExecStart=/opt/uhppoted/uhppoted-mqtt/bin/uhppoted-mqtt
...
...
```

(**WARNING: I haven't tested this so you're own for the moment**)

**Comment by YBonline** [comment]
Thanks, after I realized the problem I did edit my service file exactly as you described and that does resolve the issue, just mainly wanted to document it or have a better solution then deleting lockfiles on startup in the long term 👍

**Comment by uhppoted** [comment]
I half-suspected that was what you doing :-). Good to know it works - and thanks for letting me know! I almost think that that is probably the best solution (definitely the simplest anyway) but am going to experiment with:

- a _soft lock_, where the lock file is touched e.g. every 30s and if it hasn't been touched in e.g. the last 60s then it's not in active use
- terminating the service on too many reconnects within e.g. 30s

They're both straightforward to implement but a little finicky to test so it's going to take a week or two'ish.

Just to explain in a little more detail - the _uhppoted.lock_ file is there to prevent multiple copies of _uhppoted-mqtt_ using the same client ID (e.g. if it's already running as a service and you then start an instance from the console). It's an edge case and shouldn't happen in normal use but if it does it causes a reconnect storm which is a little catastrophic:

- it pretty much locks up the MQTT broker if you're running with a local instance of Mosquito or HiveMQ
- you're going to get your IP address banned and/or lock up the broker if you're using one of the public MQTT brokers
- you could potentially and inadvertently rack up a sizeable AWS or Google bill if you're using Greengrass or Google Core IoT

**Comment by uhppoted** [comment]
And that's a wrap ..the master branch of _uhppoted-mqtt_ implements:

- a _'soft' lockfile_ for the client connection. If the _lockfile_ exists on startup it will be ignored unless it has been updated within the last minute. This means that starting _uhppoted-mqtt_ with a leftover _lockfile_ will be slower - but it won't fail. under normal circumstances there won't be any difference - it only manifests if the system didn't shutdown cleanly.

- _disconnect monitoring_. The application will terminate if it detects too many MQTT client disconnects within the monitoring interval (defaults to 10 disconnects in 5 minutes), on the grounds that something is probably very wrong with the system and it's better not to make it worse.

Both additions are configurable in _uhppoted.conf_ and are (provisionally) enabled by default:
```
...
mqtt.softlock.enabled = true
mqtt.softlock.interval = 30s
mqtt.softlock.wait = 45s

mqtt.disconnects.interval = 5m
mqtt.disconnects.max = 10
...
```

**Comment by uhppoted** [comment]
Feature complete, pending release

**Comment by uhppoted** [comment]
Closing this - fix is release in [v0.8.2](https://github.com/uhppoted/uhppoted-mqtt/releases/tag/v0.8.2).

Feel free to reopen it if necessary :-).
